File: vn/src/main/java/org/sqlbuilder/vn/VnClass.java
package org.sqlbuilder.vn;

import org.sqlbuilder.common.Insertable;

import java.util.*;

public class VnClass implements Insertable, Comparable<VnClass>
{
	protected static final Set<VnClass> SET = new HashSet<>();

	public static Map<VnClass, Integer> MAP;

	private final String name;

	public VnClass(final String name)
	{
		this.name = name;
	}

	// A C C E S S

	public String getTag()
	{
		final int split = this.name.indexOf('-');
		return this.name.substring(split + 1);
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		VnClass that = (VnClass) o;
		return name.equals(that.name);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(name);
	}

	// O R D E R I N G

	@Override
	public int compareTo(final VnClass that)
	{
		return name.compareTo(that.name);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return this.name;
	}

	// I N S E R T

	@Override
	public String dataRow()
	{
		// id
		// name
		// tag
		return String.format("'%s','%s'", name, getTag());
	}
}

File: vn/src/main/java/org/semantikos/vn/objects/VnClass.kt
package org.semantikos.vn.objects

import org.semantikos.common.HasId
import org.semantikos.common.Insertable
import org.semantikos.common.SetCollector
import java.util.*

class VnClass private constructor(
	val name: String
) : HasId, Insertable, Comparable<VnClass> {

    val tag: String
        get() {
            val split = name.indexOf('-')
            return name.substring(split + 1)
        }

    override fun getIntId(): Int {
        return COLLECTOR.invoke(this)
    }

    // I D E N T I T Y

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val that = other as VnClass
        return name == that.name
    }

    override fun hashCode(): Int {
        return Objects.hash(name)
    }

    // O R D E R I N G

    override fun compareTo(other: VnClass): Int {
        return COMPARATOR.compare(this, other)
    }

    // T O S T R I N G

    override fun toString(): String {
        return name
    }

    // I N S E R T

    override fun dataRow(): String {
        return "'$name','$tag'"
    }

    companion object {

        val COMPARATOR: Comparator<VnClass> = compareBy { it.name }

        val COLLECTOR: SetCollector<VnClass> = SetCollector(COMPARATOR)

        fun make(name: String): VnClass {
            if (name.isEmpty()) {
                throw RuntimeException("No name")
            }

            val c = VnClass(name)
            COLLECTOR.add(c)
            return c
        }
    }
}
